eventrouter/internal/eventrouter_os.c

```diff
@@ -55,11 +69,75 @@ static void DefaultSendEvent(QueueHandle_t a_queue, void *a_event)
     }
 }
 
-static TaskHandle_t DefaultGetCurrentTaskHandle(void)
+static void DefaultReceiveEvent(ErQueueHandle_t a_queue, ErEvent_t **a_event)
+{
+    ER_ASSERT(pdTRUE == xQueueReceive(a_queue, a_event, portMAX_DELAY));
+}
+
+static void DefaultTimedReceiveEvent(ErQueueHandle_t a_queue,
+                                     ErEvent_t **a_event, int64_t a_ms)
+{
+    xQueueReceive(a_queue, a_event, pdMS_TO_TICKS(a_ms));
+}
+
+static ErTaskHandle_t DefaultGetCurrentTaskHandle(void)
 {
     return xTaskGetCurrentTaskHandle();
 }
 
+static void WaitUntilInitComplete(void)
+{
+    // FreeRTOS doesn't start tasks at creation; no delay is necessary.
+}
+#elif ER_IMPLEMENTATION == ER_IMPL_POSIX
+static void DefaultSendEvent(ErQueueHandle_t a_queue, void *a_event)
+{
+    ErQueuePushBack(a_queue, a_event);
+}
+
+static void DefaultReceiveEvent(ErQueueHandle_t a_queue, ErEvent_t **a_event)
+{
+    *a_event = ErQueuePopFront(a_queue);
+}
+
+static void DefaultTimedReceiveEvent(ErQueueHandle_t a_queue,
+                                     ErEvent_t **a_event, int64_t a_ms)
+{
+    ErQueueTimedPopFront(a_queue, a_event, a_ms);
+}
+
+static ErTaskHandle_t DefaultGetCurrentTaskHandle(void)
+{
+    return pthread_self();
+}
+
+static void WaitUntilInitComplete(void)
+{
+    while (!s_context.m_initialized)
+    {
+        pthread_mutex_lock(&s_init_gate_mutex);
+        if (!s_context.m_initialized)
+        {
+            pthread_cond_wait(&s_init_gate_cond, &s_init_gate_mutex);
+        }
+        pthread_mutex_unlock(&s_init_gate_mutex);
+    }
```

Comment:
This code works, but it forces all callers to take the lock, which is expensive.

Since `m_initialized` will almost always be true and should only transition once, it's better to check the condition and only take the lock and wait on the condition variable if that transition hasn't happened yet.

Reply:
Ah, that makes sense. Agreed.